File: scenes/plume_2d.py
#
# Simple example scene for a 2D simulation
# Simulation of a buoyant smoke density plume
#
from manta import *

# solver params
res = 64
gs = vec3(res,res,1)
s = Solver(name='main', gridSize = gs, dim=2)
s.timestep = 1.0
timings = Timings()

# prepare grids
flags = s.create(FlagGrid)
vel = s.create(MACGrid)
density = s.create(RealGrid)
pressure = s.create(RealGrid)

flags.initDomain()
flags.fillGrid()

if (GUI):
	gui = Gui()
	gui.show( True ) 
	#gui.pause()

source = s.create(Cylinder, center=gs*vec3(0.5,0.1,0.5), radius=res*0.14, z=gs*vec3(0, 0.02, 0))
	
#main loop
for t in range(400):
	if t<300:
		source.applyToGrid(grid=density, value=1)
		
	advectSemiLagrange(flags=flags, vel=vel, grid=density, order=2)    
	advectSemiLagrange(flags=flags, vel=vel, grid=vel,     order=2)
	
	setWallBcs(flags=flags, vel=vel)    
	addBuoyancy(density=density, vel=vel, gravity=vec3(0,-4e-3,0), flags=flags)
	
	solvePressure(flags=flags, vel=vel, pressure=pressure)
	setWallBcs(flags=flags, vel=vel)
	
	timings.display()    
	s.step()

File: source/tests/plume_2d.cpp
#include "manta.h"
#include "timing.h"
#include "grid.h"
#include "extforces.h"
#include "shapes.h"
#include "advection.h"
#include "pressure.h"
#include "fileio.h"

int main()
{
    int res = 64;
    auto gs = Manta::Vec3i(res, res, 1);
    auto main_solver = Manta::FluidSolver(gs, 2);
    main_solver.setTimeStep(0.5f);
    auto timings = Manta::Timings();
    timings.setParent(&main_solver);

    // prepare grids
    auto flags      = Manta::FlagGrid(&main_solver);
    auto vel        = Manta::MACGrid(&main_solver);
    auto density    = Manta::Grid<Real>(&main_solver);
    auto pressure   = Manta::Grid<Real>(&main_solver);
    
    flags.initDomain();
    flags.fillGrid();

    Manta::setOpenBound(flags, 1, "yY", Manta::FlagGrid::TypeOutflow | Manta::FlagGrid::TypeEmpty);

    auto source = Manta::Cylinder(&main_solver, Manta::Vec3(64.0f * 0.5f,
                                                            64.0f * 0.1f,
                                                            1.0f * 0.5f),
                                                            res * 0.14f, Manta::Vec3(0.0f,
                                                                                     64.0f * 0.02f,
                                                                                     0.0f));

    for (int i = 0; i < 400; i++)
    {

        if (i < 300) source.applyToGrid<Real>(&density, 1.0f);

        Manta::advectSemiLagrange(&flags, &vel, &density, 2);
        Manta::advectSemiLagrange(&flags, &vel, &vel, 2, 1.0f, 1, true, 1);
        Manta::resetOutflow(flags, nullptr, nullptr, &density);

        Manta::setWallBcs(flags, vel);
        Manta::addBuoyancy(flags, density, vel, Manta::Vec3(0.0f, -0.003f, 0.0f));

        Manta::solvePressure(vel, pressure, flags);

        //timings.display();
        main_solver.step();

        //if (i == 150) Manta::writeGridTxt("test150.txt", &density);
    }
}
